##  Ejercicio 6

In [52]:
#Importamos las librerias
import numpy as np
import pandas as pd
from matplotlib.pyplot import subplots
import matplotlib.pyplot as plt

import statsmodels.api as sm

from statsmodels.stats.outliers_influence import variance_inflation_factor as VIF
from statsmodels.stats.anova import anova_lm
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error

from ISLP import load_data
from ISLP.models import (ModelSpec as MS,summarize,poly)

In [53]:
#cargamos las base de datos
df = pd.read_csv("AsientoBebe.csv")
df.head()

Sales  CompPrice  Income  Advertising  Population  Price ShelveLoc  Age  \
0   9.50        138      73           11         276    120       Bad   42   
1  11.22        111      48           16         260     83      Good   65   
2  10.06        113      35           10         269     80    Medium   59   
3   7.40        117     100            4         466     97    Medium   55   
4   4.15        141      64            3         340    128       Bad   38   

   Education Urban   US  
0         17   Yes  Yes  
1         10   Yes  Yes  
2         12   Yes  Yes  
3         14   Yes  Yes  
4         13   Yes   No

In [54]:
#necesitamos convertir las categoricas a dummys
df = pd.get_dummies(df, columns = ['Urban', 'US'])
df = df.drop(['Urban_No','US_No'], axis=1)
df.head()

Sales  CompPrice  Income  Advertising  Population  Price ShelveLoc  Age  \
0   9.50        138      73           11         276    120       Bad   42   
1  11.22        111      48           16         260     83      Good   65   
2  10.06        113      35           10         269     80    Medium   59   
3   7.40        117     100            4         466     97    Medium   55   
4   4.15        141      64            3         340    128       Bad   38   

   Education  Urban_Yes  US_Yes  
0         17          1       1  
1         10          1       1  
2         12          1       1  
3         14          1       1  
4         13          1       0

1)

In [55]:
#Creamos X:
X = df[['Price','Urban_Yes','US_Yes']]
X.insert(0, "Ones", 1)
X.head()

Ones  Price  Urban_Yes  US_Yes
0     1    120          1       1
1     1     83          1       1
2     1     80          1       1
3     1     97          1       1
4     1    128          1       0

In [56]:
#Creamos Y:
y = df['Sales']
y.head()

0     9.50
1    11.22
2    10.06
3     7.40
4     4.15
Name: Sales, dtype: float64

In [57]:
#Estimamos 
model = sm.OLS(y, X.astype(float))
results = model.fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  Sales   R-squared:                       0.239
Model:                            OLS   Adj. R-squared:                  0.234
Method:                 Least Squares   F-statistic:                     41.52
Date:                Fri, 19 Jan 2024   Prob (F-statistic):           2.39e-23
Time:                        17:52:01   Log-Likelihood:                -927.66
No. Observations:                 400   AIC:                             1863.
Df Residuals:                     396   BIC:                             1879.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Ones          13.0435      0.651     20.036      0.000      11.764      14.323
Price         -0.0545      0.005    -10.389      0.000      -0.065      -0.044
Urban_Yes     -0.0219      0.272     -0.081      0.936      -0.556       0.512
US_Yes         1.2006      0.259      4.635      0.000       0.691       1.710
==============================================================================
Omnibus:                        0.676   Durbin-Watson:                   1.912
Prob(Omnibus):                  0.713   Jarque-Bera (JB):                0.758
Skew:                           0.093   Prob(JB):                        0.684
Kurtosis:                       2.897   Cond. No.                         628.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

2. Los coeficientes obtenidos ofrecen las siguientes interpretaciones:

    El intercepto es 13.04, indicando que el nivel de ventas "base" es de 13 asientos.

    Se observa un efecto negativo del precio de los asientos sobre las ventas. Un incremento de un dólar conlleva a una reducción de 0.05 productos vendidos.

    El efecto de que la localidad sea urbana sobre las ventas es negativo, aunque no es estadísticamente significativo

    La ubicación en los Estados Unidos tiene un impacto positivo en las ventas. Concretamente, se traduce en un aumento promedio de 1.2 asientos vendidos.

3. Se puede rechazar la hipotesis nula de H0 para el intercepto, el precio y la dummy de Estados Unidos

4.

In [58]:
X2 = df[['Price','US_Yes']]
X2.insert(0, "Ones", 1)

model = sm.OLS(y, X2.astype(float))
results = model.fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  Sales   R-squared:                       0.239
Model:                            OLS   Adj. R-squared:                  0.235
Method:                 Least Squares   F-statistic:                     62.43
Date:                Fri, 19 Jan 2024   Prob (F-statistic):           2.66e-24
Time:                        17:52:01   Log-Likelihood:                -927.66
No. Observations:                 400   AIC:                             1861.
Df Residuals:                     397   BIC:                             1873.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Ones          13.0308      0.631     20.652      0.000      11.790      14.271
Price         -0.0545      0.005    -10.416      0.000      -0.065      -0.044
US_Yes         1.1996      0.258      4.641      0.000       0.692       1.708
==============================================================================
Omnibus:                        0.666   Durbin-Watson:                   1.912
Prob(Omnibus):                  0.717   Jarque-Bera (JB):                0.749
Skew:                           0.092   Prob(JB):                        0.688
Kurtosis:                       2.895   Cond. No.                         607.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

5. Guiándonos del R cuadrado ajustado, el modelo que mejor ajusta a los datos es el segundo

6.

In [64]:
X_train, X_test, y_train, y_test = train_test_split(X,y, random_state=100, test_size=0.25, shuffle=True) 
X2_train, X2_test, y_train, y_test = train_test_split(X2,y, random_state=100, test_size=0.25, shuffle=True) 

In [65]:
for i in [X_train,X_test,X2_train,X2_test]:
  i = i.astype(float)

In [66]:
#Para el primer modelo:
model = sm.OLS(y_train, X_train)
results = model.fit()

#predicciones:
train_predictions = results.get_prediction(X_train)
y_train_pred = train_predictions.predicted_mean

test_predictions = results.get_prediction(X_test)
y_test_pred = test_predictions.predicted_mean

#MSE:
mse_train = mean_squared_error(y_train, y_train_pred)
mse_test = mean_squared_error(y_test, y_test_pred)

In [67]:
#Para el segundo modelo:
model2 = sm.OLS(y_train, X2_train)
results2 = model2.fit()

#predicciones:
train2_predictions = results2.get_prediction(X2_train)
y2_train_pred = train2_predictions.predicted_mean

test2_predictions = results2.get_prediction(X2_test)
y2_test_pred = test2_predictions.predicted_mean

#MSE:
mse_train = mean_squared_error(y_train, y2_train_pred)
mse_test = mean_squared_error(y_test, y2_test_pred)

In [69]:
print('===Primer Modelo===')
print(f'MSE en el conjunto de entrenamiento: {mse_train:.3f}')
print(f'MSE en el conjunto de prueba: {mse_test:.3f}')
print("\n")
print('===Segundo Modelo===')
print(f'MSE en el conjunto de entrenamiento: {mse2_train:.3f}')
print(f'MSE en el conjunto de prueba: {mse2_test:.3f}')

===Primer Modelo===
MSE en el conjunto de entrenamiento: 5.962
MSE en el conjunto de prueba: 6.499


===Segundo Modelo===
MSE en el conjunto de entrenamiento: 5.903
MSE en el conjunto de prueba: 6.510


Ambos modelos presentan MSE similares en el conjunto de prueba, pero el Primer Modelo exhibe un ligeramente menor MSE en el test. Considerando únicamente el rendimiento en el test, escogemos el primer modelo, pues tiene realiza mejores predicciones